File: web/src/physics/dynamics.ts
/** Numeric body-rate quadrotor core. State layout matches the Python 17-field state. */
import type { VehicleConfig, Vector3, Vector4 } from "../contracts/types";

export const clamp = (value:number, low:number, high:number) => Math.max(low,Math.min(value,high));
export function normalize(q: readonly number[]): Vector4 {
  const norm=Math.hypot(...q);
  if(q.length!==4 || !Number.isFinite(norm) || norm<1e-12) throw new Error("Invalid quaternion");
  return q.map(x=>x/norm) as Vector4;
}
export function multiply(q:readonly number[],r:readonly number[]):Vector4 {
  const [w,x,y,z]=q as Vector4,[a,b,c,d]=r as Vector4;
  return [w*a-x*b-y*c-z*d,w*b+x*a+y*d-z*c,w*c-x*d+y*a+z*b,w*d+x*c-y*b+z*a];
}
export function cross(a:readonly number[],b:readonly number[]):Vector3 {
  return [a[1]!*b[2]!-a[2]!*b[1]!,a[2]!*b[0]!-a[0]!*b[2]!,a[0]!*b[1]!-a[1]!*b[0]!];
}
export function rotation(q:readonly number[]):number[][] {
  const [w,x,y,z]=normalize(q);
  return [[1-2*(y*y+z*z),2*(x*y-w*z),2*(x*z+w*y)],
    [2*(x*y+w*z),1-2*(x*x+z*z),2*(y*z-w*x)],
    [2*(x*z-w*y),2*(y*z+w*x),1-2*(x*x+y*y)]];
}

export class Quadrotor {
  readonly dt:number;
  readonly b:number;
  constructor(readonly config:VehicleConfig){this.dt=1/config.physics_hz;this.b=config.arm_length_m/Math.sqrt(2);}

  torque(f:readonly number[]):Vector3 {
    const [a,b,c,d]=f as Vector4,k=this.config.reaction_torque_per_thrust_m;
    return [this.b*(a+b-c-d),this.b*(-a+b+c-d),k*(a-b+c-d)];
  }

  derivative(state:readonly number[],commands:readonly number[]):number[] {
    const q=state.slice(3,7),v=state.slice(7,10),omega=state.slice(10,13),motors=state.slice(13);
    const result=new Array<number>(17).fill(0),matrix=rotation(q),config=this.config;
    const torque=this.torque(motors),gyro=cross(omega,omega.map((x,i)=>x*config.inertia_kgm2[i]!));
    const dq=multiply(q,[0,...omega]);
    for(let i=0;i<3;i++) {
      result[i]=v[i]!;
      result[7+i]=matrix[i]![2]!*motors.reduce((a,b)=>a+b,0)/config.mass_kg-config.linear_drag_kgps*v[i]!/config.mass_kg;
      result[10+i]=(torque[i]!-gyro[i]!)/config.inertia_kgm2[i]!;
    }
    result[9]=result[9]!-config.gravity_mps2;
    for(let i=0;i<4;i++){result[3+i]=.5*dq[i]!;result[13+i]=(commands[i]!-motors[i]!)/config.motor_time_constant_s;}
    return result;
  }

  step(state:readonly number[],commands:readonly number[],h=this.dt):number[] {
    if(state.length!==17 || commands.length!==4 || ![...state,...commands,h].every(Number.isFinite) || h<=0 || h>this.dt+1e-12) throw new Error("Invalid physics input");
    if(commands.some(x=>x<0 || x>this.config.rotor_max_thrust_n)) throw new Error("Motor command out of bounds");
    const add=(delta:number[],scale:number)=>state.map((x,i)=>x+scale*delta[i]!);
    const a=this.derivative(state,commands),b=this.derivative(add(a,h/2),commands);
    const c=this.derivative(add(b,h/2),commands),d=this.derivative(add(c,h),commands);
    const next=state.map((x,i)=>x+h*(a[i]!+2*b[i]!+2*c[i]!+d[i]!)/6);
    if(!next.every(Number.isFinite)) throw new Error("Nonfinite RK4 state");
    next.splice(3,4,...normalize(next.slice(3,7)));
    const tolerance=this.config.motor_roundoff_tolerance_n,max=this.config.rotor_max_thrust_n;
    for(let i=13;i<17;i++){if(next[i]!< -tolerance || next[i]!>max+tolerance) throw new Error("Motor state exceeded physical bounds");next[i]=clamp(next[i]!,0,max);}
    return next;
  }
}

export function actionCommand(action:readonly number[],config:VehicleConfig):{action:Vector4;collective:number;rates:Vector3} {
  if(action.length!==4 || !action.every(Number.isFinite)) throw new Error("Action must contain four finite numbers");
  const u=action.map(x=>clamp(x,-1,1)) as Vector4;
  return {action:u,collective:(u[0]+1)*config.action.max_collective_n/2,
    rates:u.slice(1).map((x,i)=>x*config.action.max_body_rates_radps[i]!) as Vector3};
}

export function motorCommand(state:readonly number[],collective:number,rates:readonly number[],physics:Quadrotor):{motors:Vector4;scale:number} {
  if(state.length!==17 || rates.length!==3 || ![...state,...rates,collective].every(Number.isFinite) || collective<0 || collective>4*physics.config.rotor_max_thrust_n) throw new Error("Invalid controller input");
  const config=physics.config,omega=state.slice(10,13);
  const gyro=cross(omega,omega.map((x,i)=>x*config.inertia_kgm2[i]!));
  const [tx,ty,tz]=omega.map((x,i)=>config.rate_gains[i]!*(rates[i]!-x)+gyro[i]!) as Vector3;
  const b=4*physics.b,k=4*config.reaction_torque_per_thrust_m,center=collective/4;
  const delta=[tx/b-ty/b+tz/k,tx/b+ty/b-tz/k,-tx/b+ty/b+tz/k,-tx/b-ty/b-tz/k];
  let scale=1;
  for(const d of delta){if(d>0)scale=Math.min(scale,(config.rotor_max_thrust_n-center)/d);else if(d<0)scale=Math.min(scale,center/-d);}
  scale=clamp(scale,0,1);
  return {motors:delta.map(d=>clamp(center+scale*d,0,config.rotor_max_thrust_n)) as Vector4,scale};
}

File: training/physics/dynamics.py
"""17-state quadrotor, held motor commands and coupled fixed-step RK4."""

from __future__ import annotations

import numpy as np

from training.physics.quaternion import multiply, normalize, rotation


class Quadrotor:
    def __init__(self, config):
        self.config = config
        self.mass = config["mass_kg"]
        self.gravity = config["gravity_mps2"]
        self.inertia = np.array(config["inertia_kgm2"])
        self.b = config["arm_length_m"] / np.sqrt(2)
        self.kappa = config["reaction_torque_per_thrust_m"]
        self.fmax = config["rotor_max_thrust_n"]
        self.dt = 1 / config["physics_hz"]

    def torque(self, f):
        f0, f1, f2, f3 = f
        return np.array(
            [
                self.b * (f0 + f1 - f2 - f3),
                self.b * (-f0 + f1 + f2 - f3),
                self.kappa * (f0 - f1 + f2 - f3),
            ]
        )

    def derivative(self, state, motor_command):
        q, v, omega, motors = state[3:7], state[7:10], state[10:13], state[13:17]
        result = np.empty(17)
        result[:3] = v
        result[3:7] = 0.5 * multiply(q, [0, *omega])
        result[7:10] = rotation(q)[:, 2] * np.sum(motors) / self.mass
        result[7:10] -= self.config["linear_drag_kgps"] * v / self.mass
        result[9] -= self.gravity
        result[10:13] = (self.torque(motors) - np.cross(omega, self.inertia * omega)) / self.inertia
        result[13:17] = (motor_command - motors) / self.config["motor_time_constant_s"]
        return result

    def step(self, state, motor_command, dt=None):
        h = self.dt if dt is None else dt
        k1 = self.derivative(state, motor_command)
        k2 = self.derivative(state + h * k1 / 2, motor_command)
        k3 = self.derivative(state + h * k2 / 2, motor_command)
        k4 = self.derivative(state + h * k3, motor_command)
        result = state + h * (k1 + 2 * k2 + 2 * k3 + k4) / 6
        if not np.all(np.isfinite(result)):
            raise FloatingPointError("Nonfinite RK4 state")
        result[3:7] = normalize(result[3:7])
        tolerance = self.config["motor_roundoff_tolerance_n"]
        if np.any(result[13:] < -tolerance) or np.any(result[13:] > self.fmax + tolerance):
            raise FloatingPointError("Motor state exceeded physical bounds")
        result[13:] = np.clip(result[13:], 0, self.fmax)
        return result


def state_array(state):
    return np.array(
        [
            *state["position_m"],
            *state["quaternion_wxyz"],
            *state["velocity_world_mps"],
            *state["omega_body_radps"],
            *state["motor_thrust_n"],
        ],
        dtype=np.float64,
    )


def state_json(state):
    return {
        "position_m": state[:3].tolist(),
        "quaternion_wxyz": state[3:7].tolist(),
        "velocity_world_mps": state[7:10].tolist(),
        "omega_body_radps": state[10:13].tolist(),
        "motor_thrust_n": state[13:].tolist(),
    }
